radix-substate-store-interface/src/interface.rs

```diff
@@ -215,22 +240,375 @@ pub trait SubstateDatabase {
+        node_id: impl AsRef<NodeId>,
+        partition_number: PartitionNumber,
+        substate_key: impl ResolvableSubstateKey<'a>,
+    ) -> Option<Vec<u8>> {
+        self.get_raw_substate_by_db_key(
+            &db_partition_key(node_id, partition_number),
+            &db_sort_key(substate_key),
+        )
+    }
+
+    /// Gets the substate's value, if it exists, and returns it decoded as `V`.
+    ///
+    /// # Panics
+    /// This method panics if:
+    /// * There is an error decoding the value into the `V`.
+    ///
+    /// # Example use:
+    /// ```ignore
+    /// let type_info_substate = db.get_substate::<TypeInfoSubstate>(
+    ///     PACKAGE_PACKAGE,
+    ///     TYPE_INFO_FIELD_PARTITION,
+    ///     TypeInfoField::TypeInfo,
+    /// )?;
+    /// ```
+    fn get_substate<'a, V: ScryptoDecode>(
+        &self,
+        node_id: impl AsRef<NodeId>,
+        partition_number: PartitionNumber,
+        substate_key: impl ResolvableSubstateKey<'a>,
+    ) -> Option<V> {
+        let raw = self.get_raw_substate(node_id, partition_number, substate_key)?;
+        Some(decode_value(&raw))
+    }
+
+    // ------------------------------------------------------------------------------------
+    // LIST RAW
+    // ------------------------------------------------------------------------------------
+
+    /// Returns an iterator of the substates of a partition from an inclusive start cursor.
+    ///
+    /// The iterator returns raw keys and values.
+    ///
+    /// Pass `None::<SubstateKey>` as the cursor to iterate from the start of the partition.
+    #[inline]
+    fn list_raw_values<'a>(
+        &self,
+        node_id: impl AsRef<NodeId>,
+        partition_number: PartitionNumber,
+        from_substate_key_inclusive: impl ResolvableOptionalSubstateKey<'a>,
+    ) -> Box<dyn Iterator<Item = (DbSortKey, Vec<u8>)> + '_> {
+        self.list_raw_values_from_db_key(
+            &db_partition_key(node_id, partition_number),
+            optional_db_sort_key(from_substate_key_inclusive).as_ref(),
+        )
+    }
+
+    // ------------------------------------------------------------------------------------
+    // LIST KINDED PARTITIONS (OF A KNOWN BUT GENERIC KIND)
+    // ------------------------------------------------------------------------------------
+    // NOTE: There is not `list_kinded_entries` because mapping of the key requires knowing
+    // the specific kind of the substate key.
+    // ------------------------------------------------------------------------------------
+
+    /// Returns an iterator of the substates of a partition from an inclusive start cursor.
+    ///
+    /// The iterator returns `K` and the raw value for each substate.
+    /// The caller must specify `K` as [`FieldKey`], [`MapKey`] or [`SortedKey`].
+    ///
+    /// Pass `None::<SubstateKey>` as the cursor to iterate from the start of the partition.
+    fn list_kinded_raw_values<'a, K: SubstateKeyContent>(
```

Comment:
the word "kinded" hurts my eyes, but I am not the native speaker here, and I also understand that you need some word to mean "with known kind of collection" (and I also don't have better ideas).